File: sdk/src/clients/compute.ts
/**
 * Compute-to-Data Client
 *
 * Privacy-preserving computation on data inspired by Ocean Protocol.
 * Enables running algorithms on private data without exposing the underlying content.
 */

import { HttpClient } from '../utils/client';

// ============================================================================
// Types
// ============================================================================

/** Types of data assets */
export type DataAssetType =
  | 'contract'
  | 'entry'
  | 'entry_set'
  | 'dispute'
  | 'settlement'
  | 'custom';

/** Types of compute algorithms */
export type ComputeAlgorithmType =
  | 'statistical'
  | 'classification'
  | 'matching'
  | 'verification'
  | 'extraction'
  | 'analysis'
  | 'custom';

/** Access levels for data assets */
export type AccessLevel =
  | 'none'
  | 'metadata_only'
  | 'aggregate_only'
  | 'compute_only'
  | 'full_compute'
  | 'full_access';

/** Status of compute jobs */
export type JobStatus =
  | 'pending'
  | 'queued'
  | 'running'
  | 'completed'
  | 'failed'
  | 'cancelled'
  | 'timeout';

/** Privacy levels for results */
export type PrivacyLevel = 'raw' | 'anonymized' | 'aggregated' | 'differential';

/** Types of compute events */
export type ComputeEventType =
  | 'asset_registered'
  | 'asset_updated'
  | 'asset_revoked'
  | 'algorithm_registered'
  | 'access_granted'
  | 'access_revoked'
  | 'job_submitted'
  | 'job_started'
  | 'job_completed'
  | 'job_failed'
  | 'result_retrieved';

/** A registered data asset */
export interface DataAsset {
  asset_id: string;
  asset_type: DataAssetType;
  owner: string;
  name: string;
  description?: string;
  content_hash?: string;
  entry_count: number;
  allowed_algorithms: string[];
  allowed_compute_providers: string[];
  privacy_level: PrivacyLevel;
  min_aggregation_size: number;
  metadata: Record<string, unknown>;
  created_at: string;
  updated_at: string;
  is_active: boolean;
}

/** A registered compute algorithm */
export interface ComputeAlgorithm {
  algorithm_id: string;
  algorithm_type: ComputeAlgorithmType;
  name: string;
  description?: string;
  author?: string;
  code_hash?: string;
  version: string;
  input_schema: Record<string, unknown>;
  output_schema: Record<string, unknown>;
  privacy_preserving: boolean;
  supports_differential_privacy: boolean;
  min_input_size: number;
  audited: boolean;
  created_at: string;
  is_active: boolean;
}

/** An access token for compute operations */
export interface AccessToken {
  token_id: string;
  asset_id: string;
  grantee: string;
  access_level: AccessLevel;
  allowed_algorithms: string[];
  max_uses: number;
  uses_remaining: number;
  expires_at: string;
  created_at: string;
  revoked: boolean;
  valid: boolean;
}

/** A compute job */
export interface ComputeJob {
  job_id: string;
  asset_id: string;
  algorithm_id: string;
  requester: string;
  access_token_id: string;
  parameters: Record<string, unknown>;
  privacy_level: PrivacyLevel;
  status: JobStatus;
  started_at?: string;
  completed_at?: string;
  result_hash?: string;
  error_message?: string;
  created_at: string;
  compute_time_ms?: number;
}

/** Result from a compute job */
export interface ComputeResult {
  result_id: string;
  job_id: string;
  asset_id: string;
  algorithm_id: string;
  requester: string;
  data: Record<string, unknown>;
  privacy_level: PrivacyLevel;
  record_count: number;
  result_hash: string;
  created_at: string;
  expires_at?: string;
}

/** A compute event */
export interface ComputeEvent {
  event_id: string;
  event_type: ComputeEventType;
  timestamp: string;
  data: Record<string, unknown>;
}

/** Request to register a data asset */
export interface RegisterAssetRequest {
  asset_type: DataAssetType;
  owner: string;
  name: string;
  data: Record<string, unknown>[];
  description?: string;
  entry_refs?: Array<{ block_index: number; entry_index: number }>;
  allowed_algorithms?: string[];
  allowed_compute_providers?: string[];
  privacy_level?: PrivacyLevel;
  min_aggregation_size?: number;
  metadata?: Record<string, unknown>;
}

/** Request to update a data asset */
export interface UpdateAssetRequest {
  owner: string;
  allowed_algorithms?: string[];
  allowed_compute_providers?: string[];
  privacy_level?: PrivacyLevel;
  min_aggregation_size?: number;
  metadata?: Record<string, unknown>;
}

/** Request to grant access */
export interface GrantAccessRequest {
  asset_id: string;
  owner: string;
  grantee: string;
  access_level: AccessLevel;
  allowed_algorithms?: string[];
  max_uses?: number;
  expires_in_hours?: number;
}

/** Request to register an algorithm */
export interface RegisterAlgorithmRequest {
  algorithm_type: ComputeAlgorithmType;
  name: string;
  code_hash: string;
  author?: string;
  description?: string;
  input_schema?: Record<string, unknown>;
  output_schema?: Record<string, unknown>;
  privacy_preserving?: boolean;
}

/** Request to submit a compute job */
export interface SubmitJobRequest {
  asset_id: string;
  algorithm_id: string;
  access_token_id: string;
  requester: string;
  parameters?: Record<string, unknown>;
  privacy_level?: PrivacyLevel;
}

/** Response listing assets */
export interface AssetsResponse {
  count: number;
  assets: DataAsset[];
}

/** Response listing algorithms */
export interface AlgorithmsResponse {
  count: number;
  algorithms: ComputeAlgorithm[];
}

/** Response listing access tokens */
export interface AccessTokensResponse {
  count: number;
  tokens: AccessToken[];
}

/** Response listing jobs */
export interface JobsResponse {
  count: number;
  jobs: ComputeJob[];
}

/** Response listing events */
export interface ComputeEventsResponse {
  count: number;
  events: ComputeEvent[];
}

/** Compute service statistics */
export interface ComputeStatistics {
  assets: {
    total: number;
    active: number;
    by_type: Record<string, number>;
  };
  algorithms: {
    total: number;
    builtin: number;
    custom: number;
    by_type: Record<string, number>;
  };
  access_tokens: {
    total: number;
    valid: number;
  };
  jobs: {
    total: number;
    by_status: Record<string, number>;
    avg_compute_time_ms: number;
  };
  results: {
    total: number;
  };
  events: {
    total: number;
  };
}

// ============================================================================
// Client
// ============================================================================

/**
 * Client for compute-to-data operations
 *
 * @example
 * ```ts
 * // Register a data asset
 * const asset = await client.compute.registerAsset({
 *   asset_type: 'contract',
 *   owner: 'did:nlc:z6Mk...',
 *   name: 'My Contracts',
 *   data: contractData,
 *   privacy_level: 'aggregated'
 * });
 *
 * // Grant access to a compute provider
 * const token = await client.compute.grantAccess({
 *   asset_id: asset.asset_id,
 *   owner: 'did:nlc:z6Mk...',
 *   grantee: 'did:nlc:z6Mj...',
 *   access_level: 'compute_only'
 * });
 *
 * // Submit a compute job
 * const job = await client.compute.submitJob({
 *   asset_id: asset.asset_id,
 *   algorithm_id: 'builtin_count',
 *   access_token_id: token.token_id,
 *   requester: 'did:nlc:z6Mj...',
 *   parameters: { field: 'status', value: 'active' }
 * });
 *
 * // Get the result
 * const result = await client.compute.getJobResult(job.job_id, 'did:nlc:z6Mj...');
 * ```
 */
export class ComputeClient {
  constructor(private readonly http: HttpClient) {}

  // ===========================================================================
  // Data Asset Management
  // ===========================================================================

  /**
   * Register a new data asset for compute
   *
   * @param request - Asset registration options
   * @returns Registered asset info
   */
  async registerAsset(request: RegisterAssetRequest): Promise<DataAsset> {
    return this.http.post<DataAsset>('/compute/assets', request);
  }

  /**
   * Get a data asset by ID
   *
   * @param assetId - The asset ID
   * @returns Asset info (excludes actual data)
   */
  async getAsset(assetId: string): Promise<DataAsset> {
    return this.http.get<DataAsset>(`/compute/assets/${encodeURIComponent(assetId)}`);
  }

  /**
   * Update a data asset's configuration
   *
   * @param assetId - The asset ID
   * @param request - Update options
   * @returns Updated asset info
   */
  async updateAsset(assetId: string, request: UpdateAssetRequest): Promise<DataAsset> {
    return this.http.patch<DataAsset>(
      `/compute/assets/${encodeURIComponent(assetId)}`,
      request
    );
  }

  /**
   * Revoke a data asset
   *
   * @param assetId - The asset ID
   * @param owner - Asset owner (for authorization)
   * @returns Revocation result
   */
  async revokeAsset(
    assetId: string,
    owner: string
  ): Promise<{ asset_id: string; revoked: boolean }> {
    return this.http.post<{ asset_id: string; revoked: boolean }>(
      `/compute/assets/${encodeURIComponent(assetId)}/revoke`,
      { owner }
    );
  }

  /**
   * List data assets
   *
   * @param options - Filter options
   * @returns List of assets
   */
  async listAssets(options?: {
    owner?: string;
    asset_type?: DataAssetType;
    active_only?: boolean;
  }): Promise<AssetsResponse> {
    const params = new URLSearchParams();
    if (options?.owner) {
      params.set('owner', options.owner);
    }
    if (options?.asset_type) {
      params.set('asset_type', options.asset_type);
    }
    if (options?.active_only !== undefined) {
      params.set('active_only', String(options.active_only));
    }
    const query = params.toString();
    return this.http.get<AssetsResponse>(`/compute/assets${query ? `?${query}` : ''}`);
  }

  // ===========================================================================
  // Access Control
  // ===========================================================================

  /**
   * Grant access to a data asset
   *
   * @param request - Access grant options
   * @returns Access token info
   */
  async grantAccess(request: GrantAccessRequest): Promise<AccessToken> {
    return this.http.post<AccessToken>('/compute/access/grant', request);
  }

  /**
   * Revoke an access token
   *
   * @param tokenId - The token ID
   * @param owner - Asset owner (for authorization)
   * @returns Revocation result
   */
  async revokeAccess(
    tokenId: string,
    owner: string
  ): Promise<{ token_id: string; revoked: boolean }> {
    return this.http.post<{ token_id: string; revoked: boolean }>(
      `/compute/access/${encodeURIComponent(tokenId)}/revoke`,
      { owner }
    );
  }

  /**
   * List access tokens
   *
   * @param options - Filter options
   * @returns List of access tokens
   */
  async listAccessTokens(options?: {
    asset_id?: string;
    grantee?: string;
  }): Promise<AccessTokensResponse> {
    const params = new URLSearchParams();
    if (options?.asset_id) {
      params.set('asset_id', options.asset_id);
    }
    if (options?.grantee) {
      params.set('grantee', options.grantee);
    }
    const query = params.toString();
    return this.http.get<AccessTokensResponse>(`/compute/access${query ? `?${query}` : ''}`);
  }

  // ===========================================================================
  // Algorithm Management
  // ===========================================================================

  /**
   * Register a custom algorithm
   *
   * @param request - Algorithm registration options
   * @returns Registered algorithm info
   */
  async registerAlgorithm(request: RegisterAlgorithmRequest): Promise<ComputeAlgorithm> {
    return this.http.post<ComputeAlgorithm>('/compute/algorithms', request);
  }

  /**
   * Get an algorithm by ID
   *
   * @param algorithmId - The algorithm ID
   * @returns Algorithm info
   */
  async getAlgorithm(algorithmId: string): Promise<ComputeAlgorithm> {
    return this.http.get<ComputeAlgorithm>(
      `/compute/algorithms/${encodeURIComponent(algorithmId)}`
    );
  }

  /**
   * List available algorithms
   *
   * @param options - Filter options
   * @returns List of algorithms
   */
  async listAlgorithms(options?: {
    algorithm_type?: ComputeAlgorithmType;
    privacy_preserving_only?: boolean;
    audited_only?: boolean;
  }): Promise<AlgorithmsResponse> {
    const params = new URLSearchParams();
    if (options?.algorithm_type) {
      params.set('algorithm_type', options.algorithm_type);
    }
    if (options?.privacy_preserving_only !== undefined) {
      params.set('privacy_preserving_only', String(options.privacy_preserving_only));
    }
    if (options?.audited_only !== undefined) {
      params.set('audited_only', String(options.audited_only));
    }
    const query = params.toString();
    return this.http.get<AlgorithmsResponse>(`/compute/algorithms${query ? `?${query}` : ''}`);
  }

  // ===========================================================================
  // Compute Job Execution
  // ===========================================================================

  /**
   * Submit a compute job
   *
   * @param request - Job submission options
   * @returns Job info with status
   */
  async submitJob(request: SubmitJobRequest): Promise<ComputeJob> {
    return this.http.post<ComputeJob>('/compute/jobs', request);
  }

  /**
   * Get a compute job by ID
   *
   * @param jobId - The job ID
   * @returns Job info with status
   */
  async getJob(jobId: string): Promise<ComputeJob> {
    return this.http.get<ComputeJob>(`/compute/jobs/${encodeURIComponent(jobId)}`);
  }

  /**
   * Get the result of a completed compute job
   *
   * @param jobId - The job ID
   * @param requester - Requester DID (for authorization)
   * @returns Privacy-filtered result
   */
  async getJobResult(jobId: string, requester: string): Promise<ComputeResult> {
    return this.http.get<ComputeResult>(
      `/compute/jobs/${encodeURIComponent(jobId)}/result?requester=${encodeURIComponent(requester)}`
    );
  }

  /**
   * List compute jobs
   *
   * @param options - Filter options
   * @returns List of jobs
   */
  async listJobs(options?: {
    requester?: string;
    asset_id?: string;
    status?: JobStatus;
  }): Promise<JobsResponse> {
    const params = new URLSearchParams();
    if (options?.requester) {
      params.set('requester', options.requester);
    }
    if (options?.asset_id) {
      params.set('asset_id', options.asset_id);
    }
    if (options?.status) {
      params.set('status', options.status);
    }
    const query = params.toString();
    return this.http.get<JobsResponse>(`/compute/jobs${query ? `?${query}` : ''}`);
  }

  // ===========================================================================
  // Statistics and Events
  // ===========================================================================

  /**
   * Get compute service statistics
   *
   * @returns Comprehensive statistics
   */
  async getStatistics(): Promise<ComputeStatistics> {
    return this.http.get<ComputeStatistics>('/compute/statistics');
  }

  /**
   * Get compute event log
   *
   * @param options - Filter options
   * @returns List of events
   */
  async getEvents(options?: {
    limit?: number;
    event_type?: ComputeEventType;
  }): Promise<ComputeEventsResponse> {
    const params = new URLSearchParams();
    if (options?.limit !== undefined) {
      params.set('limit', String(options.limit));
    }
    if (options?.event_type) {
      params.set('event_type', options.event_type);
    }
    const query = params.toString();
    return this.http.get<ComputeEventsResponse>(`/compute/events${query ? `?${query}` : ''}`);
  }

  // ===========================================================================
  // Supported Types
  // ===========================================================================

  /**
   * Get supported data asset types
   *
   * @returns List of supported asset types
   */
  async getAssetTypes(): Promise<{ types: DataAssetType[] }> {
    return this.http.get<{ types: DataAssetType[] }>('/compute/types/assets');
  }

  /**
   * Get supported algorithm types
   *
   * @returns List of supported algorithm types
   */
  async getAlgorithmTypes(): Promise<{ types: ComputeAlgorithmType[] }> {
    return this.http.get<{ types: ComputeAlgorithmType[] }>('/compute/types/algorithms');
  }

  /**
   * Get supported access levels
   *
   * @returns List of supported access levels
   */
  async getAccessLevels(): Promise<{ levels: AccessLevel[] }> {
    return this.http.get<{ levels: AccessLevel[] }>('/compute/types/access_levels');
  }

  /**
   * Get supported privacy levels
   *
   * @returns List of supported privacy levels
   */
  async getPrivacyLevels(): Promise<{ levels: PrivacyLevel[] }> {
    return this.http.get<{ levels: PrivacyLevel[] }>('/compute/types/privacy_levels');
  }

  /**
   * Get supported job statuses
   *
   * @returns List of supported job statuses
   */
  async getJobStatuses(): Promise<{ statuses: JobStatus[] }> {
    return this.http.get<{ statuses: JobStatus[] }>('/compute/types/job_statuses');
  }
}

File: src/api/compute.py
"""
NatLangChain - Compute-to-Data API Blueprint

REST API endpoints for privacy-preserving computation on data.
Inspired by Ocean Protocol's compute-to-data paradigm.

Provides access to:
- Register data assets for compute
- Manage access tokens
- Submit and monitor compute jobs
- Retrieve privacy-filtered results
- List available algorithms
"""

from flask import Blueprint, jsonify, request

from .state import managers

compute_bp = Blueprint("compute", __name__)


# =============================================================================
# Data Asset Management Endpoints
# =============================================================================


@compute_bp.route("/compute/assets", methods=["POST"])
def register_asset():
    """
    Register a new data asset for compute.

    Request body:
        {
            "asset_type": "contract",           // contract, entry, entry_set, dispute, settlement, custom
            "owner": "did:nlc:...",             // Owner DID
            "name": "My Dataset",
            "data": [...],                      // The actual data records
            "description": "...",               // Optional
            "entry_refs": [...],                // Optional entry references
            "allowed_algorithms": [...],        // Optional algorithm IDs
            "allowed_compute_providers": [...], // Optional DIDs allowed to compute
            "privacy_level": "aggregated",      // raw, anonymized, aggregated, differential
            "min_aggregation_size": 5,          // Optional, default 5
            "metadata": {...}                   // Optional
        }

    Returns:
        Asset info with asset_id
    """
    if not managers.compute_service:
        return jsonify({"error": "Compute service not initialized"}), 503

    data = request.get_json() or {}

    required = ["asset_type", "owner", "name", "data"]
    missing = [f for f in required if f not in data]
    if missing:
        return jsonify({"error": f"Missing required fields: {missing}"}), 400

    # Parse asset type
    from compute_to_data import DataAssetType, PrivacyLevel

    try:
        asset_type = DataAssetType(data["asset_type"])
    except ValueError:
        return jsonify({"error": f"Invalid asset_type: {data['asset_type']}"}), 400

    # Parse privacy level
    privacy_level = PrivacyLevel.AGGREGATED
    if data.get("privacy_level"):
        try:
            privacy_level = PrivacyLevel(data["privacy_level"])
        except ValueError:
            return jsonify({"error": f"Invalid privacy_level: {data['privacy_level']}"}), 400

    success, result = managers.compute_service.register_asset(
        asset_type=asset_type,
        owner=data["owner"],
        name=data["name"],
        data=data["data"],
        description=data.get("description"),
        entry_refs=data.get("entry_refs"),
        allowed_algorithms=data.get("allowed_algorithms"),
        allowed_compute_providers=data.get("allowed_compute_providers"),
        privacy_level=privacy_level,
        min_aggregation_size=data.get("min_aggregation_size", 5),
        metadata=data.get("metadata"),
    )

    if success:
        return jsonify(result), 201
    else:
        return jsonify(result), 400


@compute_bp.route("/compute/assets/<asset_id>", methods=["GET"])
def get_asset(asset_id):
    """
    Get a data asset by ID.

    Path params:
        asset_id: The asset ID

    Returns:
        Asset info (excludes actual data)
    """
    if not managers.compute_service:
        return jsonify({"error": "Compute service not initialized"}), 503

    asset = managers.compute_service.get_asset(asset_id)
    if not asset:
        return jsonify({"error": "Asset not found"}), 404

    return jsonify(asset.to_dict())


@compute_bp.route("/compute/assets/<asset_id>", methods=["PATCH"])
def update_asset(asset_id):
    """
    Update a data asset's configuration.

    Path params:
        asset_id: The asset ID

    Request body:
        {
            "owner": "did:nlc:...",               // Required for auth
            "allowed_algorithms": [...],         // Optional
            "allowed_compute_providers": [...],  // Optional
            "privacy_level": "...",              // Optional
            "min_aggregation_size": 10,          // Optional
            "metadata": {...}                    // Optional
        }

    Returns:
        Updated asset info
    """
    if not managers.compute_service:
        return jsonify({"error": "Compute service not initialized"}), 503

    data = request.get_json() or {}

    if not data.get("owner"):
        return jsonify({"error": "owner is required for authorization"}), 400

    success, result = managers.compute_service.update_asset(asset_id, data["owner"], data)

    if success:
        return jsonify(result)
    else:
        return jsonify(result), 400


@compute_bp.route("/compute/assets/<asset_id>/revoke", methods=["POST"])
def revoke_asset(asset_id):
    """
    Revoke a data asset (disable for compute).

    Path params:
        asset_id: The asset ID

    Request body:
        {
            "owner": "did:nlc:..."  // Required for auth
        }

    Returns:
        Revocation result
    """
    if not managers.compute_service:
        return jsonify({"error": "Compute service not initialized"}), 503

    data = request.get_json() or {}

    if not data.get("owner"):
        return jsonify({"error": "owner is required for authorization"}), 400

    success, result = managers.compute_service.revoke_asset(asset_id, data["owner"])

    if success:
        return jsonify(result)
    else:
        return jsonify(result), 400


@compute_bp.route("/compute/assets", methods=["GET"])
def list_assets():
    """
    List data assets.

    Query params:
        owner: Filter by owner DID (optional)
        asset_type: Filter by asset type (optional)
        active_only: Only active assets (default: true)

    Returns:
        List of assets
    """
    if not managers.compute_service:
        return jsonify({"error": "Compute service not initialized"}), 503

    from compute_to_data import DataAssetType

    owner = request.args.get("owner")
    asset_type_str = request.args.get("asset_type")
    active_only = request.args.get("active_only", "true").lower() == "true"

    asset_type = None
    if asset_type_str:
        try:
            asset_type = DataAssetType(asset_type_str)
        except ValueError:
            return jsonify({"error": f"Invalid asset_type: {asset_type_str}"}), 400

    assets = managers.compute_service.list_assets(
        owner=owner,
        asset_type=asset_type,
        active_only=active_only,
    )

    return jsonify({
        "count": len(assets),
        "assets": [a.to_dict() for a in assets],
    })


# =============================================================================
# Access Control Endpoints
# =============================================================================


@compute_bp.route("/compute/access/grant", methods=["POST"])
def grant_access():
    """
    Grant access to a data asset.

    Request body:
        {
            "asset_id": "asset_...",
            "owner": "did:nlc:...",           // Asset owner (for auth)
            "grantee": "did:nlc:...",         // DID to grant access to
            "access_level": "compute_only",   // none, metadata_only, aggregate_only, compute_only, full_compute, full_access
            "allowed_algorithms": [...],      // Optional specific algorithms
            "max_uses": 100,                  // Optional, default 100
            "expires_in_hours": 24            // Optional, default 24
        }

    Returns:
        Access token info
    """
    if not managers.compute_service:
        return jsonify({"error": "Compute service not initialized"}), 503

    data = request.get_json() or {}

    required = ["asset_id", "owner", "grantee", "access_level"]
    missing = [f for f in required if f not in data]
    if missing:
        return jsonify({"error": f"Missing required fields: {missing}"}), 400

    from compute_to_data import AccessLevel

    try:
        access_level = AccessLevel(data["access_level"])
    except ValueError:
        return jsonify({"error": f"Invalid access_level: {data['access_level']}"}), 400

    success, result = managers.compute_service.grant_access(
        asset_id=data["asset_id"],
        owner=data["owner"],
        grantee=data["grantee"],
        access_level=access_level,
        allowed_algorithms=data.get("allowed_algorithms"),
        max_uses=data.get("max_uses", 100),
        expires_in_hours=data.get("expires_in_hours", 24),
    )

    if success:
        return jsonify(result), 201
    else:
        return jsonify(result), 400


@compute_bp.route("/compute/access/<token_id>/revoke", methods=["POST"])
def revoke_access(token_id):
    """
    Revoke an access token.

    Path params:
        token_id: The token ID

    Request body:
        {
            "owner": "did:nlc:..."  // Asset owner (for auth)
        }

    Returns:
        Revocation result
    """
    if not managers.compute_service:
        return jsonify({"error": "Compute service not initialized"}), 503

    data = request.get_json() or {}

    if not data.get("owner"):
        return jsonify({"error": "owner is required for authorization"}), 400

    success, result = managers.compute_service.revoke_access(token_id, data["owner"])

    if success:
        return jsonify(result)
    else:
        return jsonify(result), 400


@compute_bp.route("/compute/access", methods=["GET"])
def list_access_tokens():
    """
    List access tokens.

    Query params:
        asset_id: Filter by asset ID (optional)
        grantee: Filter by grantee DID (optional)

    Returns:
        List of access tokens
    """
    if not managers.compute_service:
        return jsonify({"error": "Compute service not initialized"}), 503

    asset_id = request.args.get("asset_id")
    grantee = request.args.get("grantee")

    tokens = managers.compute_service.get_access_tokens(asset_id=asset_id, grantee=grantee)

    return jsonify({
        "count": len(tokens),
        "tokens": [t.to_dict() for t in tokens],
    })


# =============================================================================
# Algorithm Management Endpoints
# =============================================================================


@compute_bp.route("/compute/algorithms", methods=["POST"])
def register_algorithm():
    """
    Register a custom algorithm.

    Request body:
        {
            "algorithm_type": "statistical",   // statistical, classification, matching, verification, extraction, analysis, custom
            "name": "My Algorithm",
            "code_hash": "sha256:...",         // Hash of the algorithm code
            "author": "did:nlc:...",           // Optional
            "description": "...",              // Optional
            "input_schema": {...},             // Optional
            "output_schema": {...},            // Optional
            "privacy_preserving": true         // Optional, default true
        }

    Returns:
        Algorithm info
    """
    if not managers.compute_service:
        return jsonify({"error": "Compute service not initialized"}), 503

    data = request.get_json() or {}

    required = ["algorithm_type", "name", "code_hash"]
    missing = [f for f in required if f not in data]
    if missing:
        return jsonify({"error": f"Missing required fields: {missing}"}), 400

    from compute_to_data import ComputeAlgorithmType

    try:
        algorithm_type = ComputeAlgorithmType(data["algorithm_type"])
    except ValueError:
        return jsonify({"error": f"Invalid algorithm_type: {data['algorithm_type']}"}), 400

    success, result = managers.compute_service.register_algorithm(
        algorithm_type=algorithm_type,
        name=data["name"],
        code_hash=data["code_hash"],
        author=data.get("author"),
        description=data.get("description"),
        input_schema=data.get("input_schema"),
        output_schema=data.get("output_schema"),
        privacy_preserving=data.get("privacy_preserving", True),
    )

    if success:
        return jsonify(result), 201
    else:
        return jsonify(result), 400


@compute_bp.route("/compute/algorithms/<algorithm_id>", methods=["GET"])
def get_algorithm(algorithm_id):
    """
    Get an algorithm by ID.

    Path params:
        algorithm_id: The algorithm ID

    Returns:
        Algorithm info
    """
    if not managers.compute_service:
        return jsonify({"error": "Compute service not initialized"}), 503

    algorithm = managers.compute_service.get_algorithm(algorithm_id)
    if not algorithm:
        return jsonify({"error": "Algorithm not found"}), 404

    return jsonify(algorithm.to_dict())


@compute_bp.route("/compute/algorithms", methods=["GET"])
def list_algorithms():
    """
    List available algorithms.

    Query params:
        algorithm_type: Filter by type (optional)
        privacy_preserving_only: Only privacy-preserving (default: false)
        audited_only: Only audited algorithms (default: false)

    Returns:
        List of algorithms
    """
    if not managers.compute_service:
        return jsonify({"error": "Compute service not initialized"}), 503

    from compute_to_data import ComputeAlgorithmType

    algorithm_type_str = request.args.get("algorithm_type")
    privacy_preserving_only = request.args.get("privacy_preserving_only", "false").lower() == "true"
    audited_only = request.args.get("audited_only", "false").lower() == "true"

    algorithm_type = None
    if algorithm_type_str:
        try:
            algorithm_type = ComputeAlgorithmType(algorithm_type_str)
        except ValueError:
            return jsonify({"error": f"Invalid algorithm_type: {algorithm_type_str}"}), 400

    algorithms = managers.compute_service.list_algorithms(
        algorithm_type=algorithm_type,
        privacy_preserving_only=privacy_preserving_only,
        audited_only=audited_only,
    )

    return jsonify({
        "count": len(algorithms),
        "algorithms": [a.to_dict() for a in algorithms],
    })


# =============================================================================
# Compute Job Endpoints
# =============================================================================


@compute_bp.route("/compute/jobs", methods=["POST"])
def submit_job():
    """
    Submit a compute job.

    Request body:
        {
            "asset_id": "asset_...",
            "algorithm_id": "builtin_count",
            "access_token_id": "token_...",
            "requester": "did:nlc:...",
            "parameters": {...},               // Optional algorithm params
            "privacy_level": "aggregated"      // Optional override
        }

    Returns:
        Job info with status
    """
    if not managers.compute_service:
        return jsonify({"error": "Compute service not initialized"}), 503

    data = request.get_json() or {}

    required = ["asset_id", "algorithm_id", "access_token_id", "requester"]
    missing = [f for f in required if f not in data]
    if missing:
        return jsonify({"error": f"Missing required fields: {missing}"}), 400

    from compute_to_data import PrivacyLevel

    privacy_level = None
    if data.get("privacy_level"):
        try:
            privacy_level = PrivacyLevel(data["privacy_level"])
        except ValueError:
            return jsonify({"error": f"Invalid privacy_level: {data['privacy_level']}"}), 400

    success, result = managers.compute_service.submit_job(
        asset_id=data["asset_id"],
        algorithm_id=data["algorithm_id"],
        access_token_id=data["access_token_id"],
        requester=data["requester"],
        parameters=data.get("parameters"),
        privacy_level=privacy_level,
    )

    if success:
        return jsonify(result), 201
    else:
        return jsonify(result), 400


@compute_bp.route("/compute/jobs/<job_id>", methods=["GET"])
def get_job(job_id):
    """
    Get a compute job by ID.

    Path params:
        job_id: The job ID

    Returns:
        Job info with status
    """
    if not managers.compute_service:
        return jsonify({"error": "Compute service not initialized"}), 503

    job = managers.compute_service.get_job(job_id)
    if not job:
        return jsonify({"error": "Job not found"}), 404

    return jsonify(job.to_dict())


@compute_bp.route("/compute/jobs/<job_id>/result", methods=["GET"])
def get_job_result(job_id):
    """
    Get the result of a completed compute job.

    Path params:
        job_id: The job ID

    Query params:
        requester: Requester DID (for authorization)

    Returns:
        Privacy-filtered result
    """
    if not managers.compute_service:
        return jsonify({"error": "Compute service not initialized"}), 503

    requester = request.args.get("requester")
    if not requester:
        return jsonify({"error": "requester query parameter required"}), 400

    success, result = managers.compute_service.get_job_result(job_id, requester)

    if success:
        return jsonify(result)
    else:
        return jsonify(result), 400


@compute_bp.route("/compute/jobs", methods=["GET"])
def list_jobs():
    """
    List compute jobs.

    Query params:
        requester: Filter by requester DID (optional)
        asset_id: Filter by asset ID (optional)
        status: Filter by status (optional)

    Returns:
        List of jobs
    """
    if not managers.compute_service:
        return jsonify({"error": "Compute service not initialized"}), 503

    from compute_to_data import JobStatus

    requester = request.args.get("requester")
    asset_id = request.args.get("asset_id")
    status_str = request.args.get("status")

    status = None
    if status_str:
        try:
            status = JobStatus(status_str)
        except ValueError:
            return jsonify({"error": f"Invalid status: {status_str}"}), 400

    jobs = managers.compute_service.list_jobs(
        requester=requester,
        asset_id=asset_id,
        status=status,
    )

    return jsonify({
        "count": len(jobs),
        "jobs": [j.to_dict() for j in jobs],
    })


# =============================================================================
# Statistics and Events
# =============================================================================


@compute_bp.route("/compute/statistics", methods=["GET"])
def get_statistics():
    """
    Get compute service statistics.

    Returns:
        Comprehensive statistics
    """
    if not managers.compute_service:
        return jsonify({"error": "Compute service not initialized"}), 503

    return jsonify(managers.compute_service.get_statistics())


@compute_bp.route("/compute/events", methods=["GET"])
def get_events():
    """
    Get compute event log.

    Query params:
        limit: Maximum events to return (default: 100)
        event_type: Filter by event type (optional)

    Returns:
        List of events
    """
    if not managers.compute_service:
        return jsonify({"error": "Compute service not initialized"}), 503

    limit = request.args.get("limit", 100, type=int)
    event_type_filter = request.args.get("event_type")

    events = managers.compute_service.events

    if event_type_filter:
        events = [e for e in events if e.event_type.value == event_type_filter]

    events = events[-limit:]

    return jsonify({
        "count": len(events),
        "events": [e.to_dict() for e in reversed(events)],
    })


# =============================================================================
# Supported Types
# =============================================================================


@compute_bp.route("/compute/types/assets", methods=["GET"])
def get_asset_types():
    """
    Get supported data asset types.

    Returns:
        List of supported asset types
    """
    from compute_to_data import DataAssetType

    return jsonify({"types": [t.value for t in DataAssetType]})


@compute_bp.route("/compute/types/algorithms", methods=["GET"])
def get_algorithm_types():
    """
    Get supported algorithm types.

    Returns:
        List of supported algorithm types
    """
    from compute_to_data import ComputeAlgorithmType

    return jsonify({"types": [t.value for t in ComputeAlgorithmType]})


@compute_bp.route("/compute/types/access_levels", methods=["GET"])
def get_access_levels():
    """
    Get supported access levels.

    Returns:
        List of supported access levels
    """
    from compute_to_data import AccessLevel

    return jsonify({"levels": [level.value for level in AccessLevel]})


@compute_bp.route("/compute/types/privacy_levels", methods=["GET"])
def get_privacy_levels():
    """
    Get supported privacy levels.

    Returns:
        List of supported privacy levels
    """
    from compute_to_data import PrivacyLevel

    return jsonify({"levels": [level.value for level in PrivacyLevel]})


@compute_bp.route("/compute/types/job_statuses", methods=["GET"])
def get_job_statuses():
    """
    Get supported job statuses.

    Returns:
        List of supported job statuses
    """
    from compute_to_data import JobStatus

    return jsonify({"statuses": [s.value for s in JobStatus]})
